Library/Homebrew/style.rb

```diff
@@ -79,63 +115,98 @@ def check_style_impl(files, output_type,
-        rubocop_success = $CHILD_STATUS.success?
-      when :json
-        json, err, status =
-          Open3.capture3(cache_env, "rubocop", "--format", "json", *args)
-        # exit status of 1 just means violations were found; other numbers mean
-        # execution errors.
-        # exitstatus can also be nil if RuboCop process crashes, e.g. due to
-        # native extension problems.
-        # JSON needs to be at least 2 characters.
-        if !(0..1).cover?(status.exitstatus) || json.to_s.length < 2
-          raise "Error running `rubocop --format json #{args.join " "}`\n#{err}"
-        end
 
-        return RubocopResults.new(JSON.parse(json))
-      else
-        raise "Invalid output_type for check_style_impl: #{output_type}"
+        system cache_env, "rubocop", *args
+        $CHILD_STATUS.success?
+      when :json
+        result = system_command "rubocop", args: ["--format", "json", *args], env: cache_env
+        json = json_result!(result)
+        json["files"]
       end
+    end
 
-      return rubocop_success if files.present?
-
+    def run_shellcheck(files, output_type)
       shellcheck   = which("shellcheck")
       shellcheck ||= which("shellcheck", ENV["HOMEBREW_PATH"])
       shellcheck ||= begin
         ohai "Installing `shellcheck` for shell style checks..."
-        system HOMEBREW_BREW_FILE, "install", "shellcheck"
+        safe_system HOMEBREW_BREW_FILE, "install", "shellcheck"
         which("shellcheck") || which("shellcheck", ENV["HOMEBREW_PATH"])
       end
-      unless shellcheck
-        opoo "Could not find or install `shellcheck`! Not checking shell style."
-        return rubocop_success
-      end
-
-      shell_files = [
-        HOMEBREW_BREW_FILE,
-        *Pathname.glob("#{HOMEBREW_LIBRARY}/Homebrew/*.sh"),
-        *Pathname.glob("#{HOMEBREW_LIBRARY}/Homebrew/cmd/*.sh"),
-        *Pathname.glob("#{HOMEBREW_LIBRARY}/Homebrew/utils/*.sh"),
-      ].select(&:exist?)
-      # TODO: check, fix completions here too.
-      # TODO: consider using ShellCheck JSON output
-      shellcheck_success = system shellcheck, "--shell=bash", *shell_files
-      rubocop_success && shellcheck_success
+
+      if files.empty?
+        files = [
+          HOMEBREW_BREW_FILE,
+          # TODO: HOMEBREW_REPOSITORY/"completions/bash/brew",
+          *Pathname.glob("#{HOMEBREW_LIBRARY}/Homebrew/*.sh"),
+          *Pathname.glob("#{HOMEBREW_LIBRARY}/Homebrew/cmd/*.sh"),
+          *Pathname.glob("#{HOMEBREW_LIBRARY}/Homebrew/utils/*.sh"),
+        ]
+      end
+
+      args = ["--shell=bash", "--", *files] # TODO: Add `--enable=all` to check for more problems.
+
+      case output_type
+      when :print
+        system shellcheck, "--format=tty", *args
+        $CHILD_STATUS.success?
+      when :json
+        result = system_command shellcheck, args: ["--format=json1", *args]
+        json = json_result!(result)
+
+        # Convert to same format as RuboCop offenses.
```

Comment:
Why?

Reply:
I think it's easier to return a unified result for both RuboCop and ShellCheck offenses this way.

Reply:
Ah, unified result, makes sense 👍🏻